src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+     * for any given pair to have conflicts that are unique to that pair. GATKTool only validates individual feature
+     * dictionaries against the reference dictionary, so we need to cross-validate them against each other here
+     * before we merge them together.
+     */
+    private void validateAllSequenceDictionaries() {
+        Map<String, FeatureDataSource<VariantContext>> contigMap = new HashMap<>();
+        featureDataSources.forEach(
+            ds -> {
+                ds.getSequenceDictionary().getSequences().forEach(
+                    sourceSequence -> {
+                        String sourceName = sourceSequence.getSequenceName();
+                        FeatureDataSource<VariantContext> targetDataSource = contigMap.getOrDefault(sourceName, null);
+                        if (targetDataSource != null) {
+                            SAMSequenceDictionary targetDictionary = targetDataSource.getSequenceDictionary();
+                            SAMSequenceRecord targetSequence = targetDictionary.getSequence(sourceName);
+                            validateSequenceDictionaryRecords(
+                                    ds.getName(),
+                                    ds.getSequenceDictionary(),
+                                    sourceSequence,
+                                    targetDataSource.getName(),
+                                    targetDictionary,
+                                    targetSequence);
+                        } else {
+                            contigMap.put(sourceName, ds);
+                        }
+                    }
+                );
+            }
+        );
+    }
+
+    // Cross validate the length and md5 for a pair of sequence records and throw if they're not the same.
+    private void validateSequenceDictionaryRecords(
+            final String sourceDataSourceName,
+            final SAMSequenceDictionary sourceDictionary,
+            final SAMSequenceRecord sourceSequence,
+            final String targetDataSourceName,
+            final SAMSequenceDictionary targetDictionary,
+            final SAMSequenceRecord targetSequence)
+    {
+        if (targetSequence.getSequenceLength() != sourceSequence.getSequenceLength()) {
+            final String msg = String.format("Incompatible sequence lengths found (%s: %d) and (%s: %d)",
+                    sourceSequence.getSequenceName(),
+                    sourceSequence.getSequenceLength(),
+                    targetSequence.getSequenceName(),
+                    targetSequence.getSequenceLength());
+            throw new UserException.IncompatibleSequenceDictionaries(
+                    msg,
+                    sourceDataSourceName,
+                    sourceDictionary,
+                    targetDataSourceName,
+                    targetDictionary
+            );
+        } else if (!Objects.equals(targetSequence.getMd5(), sourceSequence.getMd5())) {
```

Comment:
I think we might want to tolerate the situation where the MD5 is filled in in one record but not the other, and just emit a warning in that case.